numba/np/ziggurat_constants.py

```diff
@@ -0,0 +1,1212 @@
+
+
+ki_double = (
```

Comment:
RE: https://github.com/numba/numba/pull/7900#pullrequestreview-909569233

I suggest making these large constant pools into `np.array()`. In the current implementation, this is a tuple so it will appear as an aggregate type in the LLVM IR and also have to adhere to Python tuple semantics. This has two effects, the first is that indexing ends up going via a massive switch table to jump to the right value (opposed to a cheap LLVM `GetElementPointer()` as is present in NumPy array indexing), and the second is that tuple indexing semantics like negative indexing and raising on invalid index have to be dealt with. The combination of these two things is probably what's causing the slow down in comparison to a relatively direct non-bounds-checked read-only NumPy array access.

Reply:
I tried making the constant pools an Numpy array, it did make the code run a bit faster, but that difference in performance was little when compared to the overall difference. Even if we remove all the extra logic and just focus on the CFFI bindings. For instance:

```python
import numpy as np

from numba import njit


@njit
def numba_standard_normal(rng):
    return rng.random(10000)


res = numba_standard_normal(np.random.default_rng(seed=230923))

%timeit _ = numba_standard_normal(np.random.default_rng(seed=230923))
# 151 µs ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%timeit _ = np.random.default_rng(seed=230923).random(10000)
# 99.8 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

exp_res = np.random.default_rng(seed=230923).random(10000)
assert np.array_equal(res, exp_res)
```

Which internally simply calls the `next_float` or `next_double` bindings, it's still slower.

Reply:
I see, if the output is 100_000 in size opposed to 10_000 does it make a difference? I guess I'll need to go and look at the machine code for both implementations. Nothing obvious jumps out from a quick skim of the code generation.

Reply:
> ```python
> import numpy as np
> 
> from numba import njit
> 
> 
> @njit
> def numba_standard_normal(rng):
>     return rng.random(10000)
> 
> 
> res = numba_standard_normal(np.random.default_rng(seed=230923))
> 
> %timeit _ = numba_standard_normal(np.random.default_rng(seed=230923))
> # 151 µs ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
> 
> %timeit _ = np.random.default_rng(seed=230923).random(10000)
> # 99.8 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
> 
> exp_res = np.random.default_rng(seed=230923).random(10000)
> assert np.array_equal(res, exp_res)
> ```

Is this example using those constants?  It looks like `numba_standard_normal` is calling `next_float` (via `NumPyRandomGeneratorType.random`), which doesn't use the constants in `numba.np.ziggurat_constants`&mdash;from what I can tell.